cli/src/complete.rs

```diff
@@ -575,6 +712,78 @@ impl JjBuilder {
     }
 }
 
+/// Functions for parsing revisions and revision ranges from the command line.
+/// Parsing is done on a best-effort basis and relies on the heuristic that
+/// most command line flags are consistent across different subcommands.
+///
+/// In some cases, this parsing will be incorrect, but it's not worth the effort
+/// to fix that. For example, if the user specifies any of the relevant flags
+/// multiple times, the parsing will pick any of the available ones, while the
+/// actual execution of the command would fail.
+mod parse {
+    fn parse_flag(candidates: &[&str], mut args: impl Iterator<Item = String>) -> Option<String> {
+        for arg in args.by_ref() {
+            // -r REV syntax
+            if candidates.contains(&arg.as_ref()) {
+                return args.next();
```

Comment:
nit: (optional) we can check if `args.next()` doesn't starts with `-`.

clap rejects `-r -foo`, which is nice for this kind of ad-hoc parsing.